### ISSUE TITLE
[Suggestion] Option to control handler lifetime

### ISSUE THREAD

**Issue by janoveh** [martinothamar/Mediator#60]
What about providing the following interfaces and use those to register handlers accordingly:
ISingletonRequestHandler<in TRequest, Tresponse>, IScopedRequestHandler<...>, ITransientRequestHandler<...>
All of them inherit from IRequestHandler<...>. Default lifetime applies if IRequestHandler<...> is used.
Alternatively, there could maybe be an attribute to override the default registration.
In a larger asp.net core web api type of backend you will typically have a mix of lifetimes for your handlers.
It would actually also be a good idea for notification handlers, in my opinion.

**Comment by martinothamar** [comment]
Interesting idea! This needs some prototyping (would need some refactoring), but it would definitely be a nice feature. I agree that it would have to be configurable across all handlers. 

Probably will be some challenges/design questions when it comes to combination of handlers and pipelinebehaviors with different lifetimes, but as long as it doesn't become too messy/unclear I think this could work

**Comment by janoveh** [comment]
Yes, I believe there will probably (?) be some challenges around handling DI scopes and injection of the IMediator service into handlers, especially in singletons, I guess, but I think they can be overcome by carefully creating scopes as needed.  But this would of course be the responsibility of the handle developer and not the Mediator source generators.
I must say I really like the work you have done on the source generator.

**Comment by pinkfloydx33** [comment]
Any status on this? 

I'm not really a fan of different marker interfaces in this case... could it be controlled at the handler-level with an attribute, similar to `MediatorOptionsAttribute` (only, with a class target and lacking the `Namespace` property)?

Most of our handlers are registered as Transient for "reasons", but we have a handful that we'd like to register as singletons because they are heavy use and don't need anything scoped... it'd be nice to be able to indicate we wanted those particular ones to be registered differently, without cluttering service registrations ourselves.

**Comment by martinothamar** [comment]
Not really, this hasn't come up much and I haven't had the need personally. I tend to go all in on `Singleton` :)

**Comment by martinothamar** [comment]
Yeah totally understandable. Would accept a PR yes. I like your idea of using an attribute on the handlers
